File: packages/python/elizaos/features/basic_capabilities/providers/world.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


async def get_world_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    room_id = message.room_id
    if not room_id:
        return ProviderResult(text="", values={"hasWorld": False}, data={})

    room = await runtime.get_room(room_id)
    if not room or not room.world_id:
        return ProviderResult(
            text="",
            values={"hasWorld": False, "roomId": str(room_id)},
            data={"roomId": str(room_id)},
        )

    world = await runtime.get_world(room.world_id)
    if not world:
        return ProviderResult(
            text="",
            values={"hasWorld": False, "roomId": str(room_id)},
            data={"roomId": str(room_id)},
        )

    sections: list[str] = []

    sections.append(f"# World: {world.name or 'Unknown'}")
    if world.metadata and world.metadata.get("description"):
        sections.append(f"\n{world.metadata['description']}")

    sections.append(f"\n## Current Room: {room.name or 'Unknown'}")
    if room.metadata and room.metadata.get("topic"):
        sections.append(f"Topic: {room.metadata['topic']}")

    member_count = 0
    if world.metadata and world.metadata.get("members"):
        member_count = len(world.metadata["members"])
    sections.append(f"\n## Members: {member_count}")

    if world.metadata and world.metadata.get("settings"):
        safe_settings = {
            k: v
            for k, v in world.metadata["settings"].items()
            if not k.lower().endswith(("key", "secret", "password", "token"))
        }
        if safe_settings:
            settings_list = [f"- {k}: {v}" for k, v in safe_settings.items()]
            sections.append("\n## Settings\n" + "\n".join(settings_list))

    context_text = "\n".join(sections)

    return ProviderResult(
        text=context_text,
        values={
            "hasWorld": True,
            "worldId": str(world.id),
            "worldName": world.name or "Unknown",
            "roomId": str(room_id),
            "roomName": room.name or "Unknown",
            "memberCount": member_count,
        },
        data={
            "world": {
                "id": str(world.id),
                "name": world.name,
                "metadata": world.metadata,
            },
            "room": {
                "id": str(room_id),
                "name": room.name,
                "metadata": room.metadata,
            },
        },
    )


world_provider = Provider(
    name="WORLD",
    description="Provides information about the current world context including settings and members",
    get=get_world_context,
    dynamic=True,
)

File: packages/rust/src/features/basic_capabilities/bootstrap/providers/world.rs
//! WORLD provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("WORLD"));

/// Provider for world/server context.
pub struct WorldProvider;

#[async_trait]
impl Provider for WorldProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let Some(room_id) = message.room_id else {
            return Ok(ProviderResult::new("")
                .with_value("hasWorld", false)
                .with_data("error", "no_room_id"));
        };

        let Some(room) = runtime.get_room(room_id).await? else {
            return Ok(ProviderResult::new("")
                .with_value("hasWorld", false)
                .with_value("roomId", room_id.to_string()));
        };

        let Some(world_id) = room.world_id else {
            return Ok(ProviderResult::new("")
                .with_value("hasWorld", false)
                .with_value("roomId", room_id.to_string()));
        };

        let Some(world) = runtime.get_world(world_id).await? else {
            return Ok(ProviderResult::new("")
                .with_value("hasWorld", false)
                .with_value("roomId", room_id.to_string()));
        };

        let mut sections = Vec::new();

        // World name and description
        let world_name = world.name.clone().unwrap_or_else(|| "Unknown".to_string());
        sections.push(format!("# World: {}", world_name));

        if let Some(desc) = world.metadata.get("description").and_then(|v| v.as_str()) {
            sections.push(format!("\n{}", desc));
        }

        // Room info
        let room_name = room.name.clone().unwrap_or_else(|| "Unknown".to_string());
        sections.push(format!("\n## Current Room: {}", room_name));

        if let Some(topic) = room.metadata.get("topic").and_then(|v| v.as_str()) {
            sections.push(format!("Topic: {}", topic));
        }

        // Member count
        let member_count = world
            .metadata
            .get("members")
            .and_then(|v| v.as_array())
            .map(|arr| arr.len())
            .unwrap_or(0);

        sections.push(format!("\n## Members: {}", member_count));

        let context_text = sections.join("\n");

        Ok(ProviderResult::new(context_text)
            .with_value("hasWorld", true)
            .with_value("worldId", world_id.to_string())
            .with_value("worldName", world_name.clone())
            .with_value("roomId", room_id.to_string())
            .with_value("roomName", room_name.clone())
            .with_value("memberCount", member_count as i64)
            .with_data(
                "world",
                serde_json::json!({
                    "id": world_id.to_string(),
                    "name": world_name
                }),
            )
            .with_data(
                "room",
                serde_json::json!({
                    "id": room_id.to_string(),
                    "name": room_name
                }),
            ))
    }
}
